From: Sebastian Ott <sebott@linux.vnet.ibm.com>
Subject: s390/pci: fmb enhancements
Patch-mainline: not yet, in process
References: FATE#319595, LTC#KRN1405

Summary:     s390/pci: fmb enhancements
Description: Implement the function type specific function measurement
             block used in new machines.

Upstream-Description:

             s390/pci: add extra padding to function measurement block

             Newer machines might use a different (larger) format for function
             measurement blocks. To ensure that we comply with the alignment
             requirement on these machines and prevent memory corruption (when
             firmware writes more data than we expect) add 16 padding bytes
             at the end of the fmb.

             Cc: stable@vger.kernel.org # v4.1+
             Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>


Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/include/asm/pci.h |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/arch/s390/include/asm/pci.h
+++ b/arch/s390/include/asm/pci.h
@@ -45,7 +45,8 @@ struct zpci_fmb {
 	u64 rpcit_ops;
 	u64 dma_rbytes;
 	u64 dma_wbytes;
-} __packed __aligned(64);
+	u64 pad[2];
+} __packed __aligned(128);
 
 enum zpci_state {
 	ZPCI_FN_STATE_RESERVED,
